pkg/codegen/pcl/binder_nodes.go

```diff
@@ -22,20 +22,57 @@ import (
 	"github.com/pulumi/pulumi/sdk/v3/go/common/util/contract"
 )
 
+// hasReferenceTo returns true if the source node has a reference to the target node.
+// In other words, if the target node is a dependency of the source node.
+func hasReferenceTo(source Node, target Node) bool {
+	for _, dep := range source.getDependencies() {
+		if dep.Name() == target.Name() {
+			return true
+		}
+	}
+	return false
+}
+
+// selfReferencingNode returns true if the given node references itself.
+func selfReferencingNode(node Node) bool {
+	return hasReferenceTo(node, node)
+}
+
+// isComponent returns true if the given node is a component.
+func isComponent(node Node) bool {
+	_, ok := node.(*Component)
+	return ok
+}
+
+// mutuallyDependantComponents returns true if the given nodes are components that reference each other.
+func mutuallyDependantComponents(a Node, b Node) bool {
+	return isComponent(a) &&
+		isComponent(b) &&
+		hasReferenceTo(a, b) &&
+		hasReferenceTo(b, a) &&
+		a.Name() != b.Name()
+}
+
 // bindNode binds a single node in a program. The node's dependencies are bound prior to the node itself; it is an
 // error for a node to depend--directly or indirectly--upon itself.
 func (b *binder) bindNode(node Node) hcl.Diagnostics {
 	if node.isBound() {
 		return nil
 	}
 	if node.isBinding() {
-		// TODO(pdg): print trace
-		rng := node.SyntaxNode().Range()
-		return hcl.Diagnostics{{
-			Severity: hcl.DiagError,
-			Summary:  "circular reference",
-			Subject:  &rng,
-		}}
+		// We encountered the same node while binding its dependencies, so we have a circular reference.
+		// However we need to make an exception for nodes of type Component when
+		// - they are not self-referencing
+		// - the circular reference is only between other nodes of type Component
+		if !isComponent(node) || selfReferencingNode(node) {
+			// TODO(pdg): print better trace
+			rng := node.SyntaxNode().Range()
+			return hcl.Diagnostics{{
+				Severity: hcl.DiagError,
+				Summary:  "circular reference",
+				Subject:  &rng,
+			}}
+		}
```

Comment:
else should we have a warning or an info diagnostic of some sort?

Reply:
No, this isn't a warning case. If it's a circular dependency between components that will just work once we've fixed codegen up to handle it.